DOCS-848: document the navigation service (#1386)
viamrobotics/docs: docs/services/navigation/_index.md

---
title: "The Navigation Service"
linkTitle: "Navigation"
description: "The Navigation Service uses GPS to autonomously navigate a rover to user defined endpoints."
type: docs

> weight: 45

Reviewer: let's put that after the ML Model service - we should probably alphabetize them. We'll also need to fix the motion one but can do that separately

Author: oh great point, done!

no_list: true

> # icon: "/services/img/icons/frame-system.svg"

Author: Done!

tags: ["navigation", "services", "base", "rover"]
# SMEs: Raymond
---

> The Navigation service is the stateful definition of Viam's [Motion service](/services/motion/)

Reviewer: ```suggestion
The Navigation service is the stateful definition of Viam's [Motion Service](/services/motion/)
```

Captialized if we're refering to the concept

Author: Done

It uses GPS to autonomously navigate a rover [base](/components/base/) to user defined endpoints called `Waypoints`.

> Once the `Waypoints` are added and the `Mode` of the service is [set to `waypoints`](#setmode), Navigation services calls begin.

Reviewer: what does Navigation services calls begin mean - does that mean "the rover/base begins to navicate to the waypoint(s)"?

Author: I think that is what it means but that explanation confused me as well. I think this might be better but it's a bit long, I could trim the top info if you think that could help?

## Configuration

{{< tabs >}}
{{% tab name="Config Builder" %}}

Navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
Click on the **Services** subtab and navigate to the **Create service** menu.
Select the type `Navigation` and enter a name for your service.

Click **Create service**:

![An example configuration for a Navigation service in the Viam app Config Builder.](img/navigation-ui-config.png)

{{% /tab %}}
{{% tab name="JSON Template" %}}

```json {class="line-numbers linkable-line-numbers"}
{
    "name": "your-navigation-service",
    "type": "navigation"
}
```

{{% /tab %}}
{{< /tabs >}}

Next, add the JSON `"attributes"` you want the service to have.
The following attributes are available for `Navigation` services:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |
| `store` | obj | **Required** | The MongoDB store ObjectID. Set this to access the backend MongoDB database. |

> | `base` | string | **Required** | The `name` of the [base](/components/base/) you have configured for the rover you are operating as with this service. |

Author: Yeah I think so?

The Navigation Service supports the following methods:

Method Name | Description
----------- | -----------
[`Mode`](#mode) | Get the mode the service is operating in.
[`SetMode`](#setmode) | Set the mode the service is operating in.
[`Location`](#location) | Get the current location of the robot.
[`Waypoints`](#waypoints) | Get an array of waypoints currently in the service's data storage.
[`AddWaypoint`](#addwaypoint) | Add a waypoint to the service's data storage.
[`RemoveWaypoint`](#removewaypoint) | Remove a waypoint from the service's data storage.

{{% alert title="Note" color="note" %}}

The following code examples assume that you have a robot configured with a `Navigation` service, and that you add the required code to connect to your robot and import any required packages at the top of your code file.
Go to your robot's **Code Sample** tab on the [Viam app](https://app.viam.com) for boilerplate code to connect to your robot.

{{% /alert %}}

### Mode

{{% alert title="Note" color="note" %}}

The `manual` mode is experimental and will have no effect.

{{% /alert %}}

Get the `Mode` the service is operating in.
There are two options for modes: `manual` and `waypoint`.

> If the mode is set to `waypoint`, the service's process for looking for waypoints that were added will begin, effectively beginning navigation.

Reviewer: ```suggestion
If you set the mode to `waypoint`, the service starts looking for added waypoints, effectively beginning navigation.
```

{{< tabs >}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.
- `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

**Returns:**

- [(Mode)](https://pkg.go.dev/go.viam.com/rdk@v0.2.49/services/navigation#Mode): The `Mode` the service is operating in.
- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk@v0.2.49/services/navigation#Service).

```go
myNav, err := navigation.FromRobot(robot, "my_nav")

// Get the Mode the service is operating in
mode, err := myNav.Mode(context.Background(), nil)
```

{{% /tab %}}
{{< /tabs >}}

### SetMode

{{% alert title="Note" color="note" %}}

The `manual` mode is experimental and will have no effect.

{{% /alert %}}

Set the `Mode` the service is operating in.
There are two options for modes: `manual` and `waypoint`.

> If the mode is set to `waypoint`, the service's process for looking for waypoints that were added will begin, effectively beginning navigation.

Reviewer: ```suggestion
If you set the mode to `waypoint`, the service starts looking for added waypoints, effectively beginning navigation.
```

Author: Done

// Set the Mode the service is operating in
mode, err := myNav.SetMode(context.Background(), "waypoint", nil)
```

{{% /tab %}}
{{< /tabs >}}

### Location

Get the current location of the robot in the navigation service.

{{< tabs >}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.
- `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

**Returns:**

- [(*geo.Point)](https://pkg.go.dev/github.com/kellydunn/golang-geo#Point): The current location of the robot in the navigation service, represented in a `Point` with latitude (lat) and longitude (lng) values.
- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk@v0.2.49/services/navigation#Service).

```go
myNav, err := navigation.FromRobot(robot, "my_nav")

// Get the current location of the robot in the navigation service
location, err := myNav.Location(context.Background(), nil)
```

{{% /tab %}}
{{< /tabs >}}

### Waypoints

> Get an array of waypoints, locations designated within a path for the robot to navigate to, currently in the service's data storage.

Reviewer: ```suggestion
Get an array of waypoints currently in the service's data storage.
These are locations designated within a path for the robot to navigate to.
```

Author: good suggestion, done

It uses GPS to autonomously navigate a rover [base](/components/base/) to user defined endpoints called `Waypoints`.
Once the `Waypoints` are added and the `Mode` of the service is [set to `waypoints`](#setmode), the service communicates with the base as it begins to navigate to the `Waypoints`.

## Configuration

{{< tabs >}}
{{% tab name="Config Builder" %}}

Navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
Click on the **Services** subtab and navigate to the **Create service** menu.
Select the type `Navigation` and enter a name for your service.

Click **Create service**:

![An example configuration for a Navigation service in the Viam app Config Builder.](img/navigation-ui-config.png)

{{% /tab %}}
{{% tab name="JSON Template" %}}

```json {class="line-numbers linkable-line-numbers"}
{
    "name": "your-navigation-service",
    "type": "navigation"
}
```

{{% /tab %}}
{{< /tabs >}}

Next, add the JSON `"attributes"` you want the service to have.
The following attributes are available for `Navigation` services:

| Name | Type | Inclusion | Description |
| ---- | ---- | --------- | ----------- |

> | `store` | obj | **Required** | The configuration to use for the Navigation data storage. There are two parameters: `type` (`"memory"` or `"mongodb"`) and `config`. <br> Default: `"store": [{ "type": "mongodb", "config": {"uri": "mongodb://127.0.0.1:27017"}]` |

Reviewer: Wait does a user need to run mongodb themselves?
